File: Arrays/Product of Array Except Self/optimal.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        
        int pre = 1;
        int post = 1;
        int len = nums.length;
        int[] out = new int[len];
        
        for(int i = 0; i < len; i++){
            
            out[i] = pre;
            
            pre *= nums[i];
            
        }
        
        for(int i = len-1; i > -1; i--){
            
            out[i] *= post;
            
            post *= nums[i];
            
        }
        
        return out;
        
    }
}

File: Arrays/Product of Array Except Self/optimal.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        pre = 1
        post = 1
        l = len(nums)
        # calc prefix from start
        
        out = [1]*l
        
        for i in range(l):
            out[i] = pre
            pre *= nums[i]
            
        for i in range(l-1,-1,-1):
            out[i] *= post
            post *= nums[i]
            
            
        return out
            
            
        
        